packages/sw-runtime-caching/src/lib/request-wrapper.js

```diff
@@ -184,11 +181,11 @@ class RequestWrapper {
       assert.isInstance({request}, Request);
     }
 
-    const clonedRequest = this.pluginCallbacks.fetchDidFail ?
+    const clonedRequest = this.plugins.has('fetchDidFail') ?
       request.clone() : null;
-    if (this.pluginCallbacks.requestWillFetch) {
-      for (let callback of this.pluginCallbacks.requestWillFetch) {
-        const returnedPromise = callback({request});
+    if (this.plugins.has('requestWillFetch')) {
+      for (let plugin of this.plugins.get('requestWillFetch')) {
```

Comment:
I believe `for ... of` used with an iterable does the right thing here, and only evaluates the entries it will iterate over once at the start of the loop.

See the hacky example at https://jsbin.com/mesudupeze/edit?js,console in which demonstrates that.